### PR TITLE
Refactor Settings and add MeterSettings 

### PR DESCRIPTION
**Pull request by rajkumar-rangaraj**
Fixes part of #662 to include `MeterSettings `

### CODE REVIEW

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            : base(source)
+        {
+            MetricExporter = ParseMetricExporter(source);
+            ConsoleExporterEnabled = source.GetBool(ConfigurationKeys.Metrics.ConsoleExporterEnabled) ?? false;
+
+            var instrumentations = new Dictionary<string, MeterInstrumentation>();
+            var enabledInstrumentations = source.GetString(ConfigurationKeys.Metrics.Instrumentations);
+            if (enabledInstrumentations != null)
+            {
+                foreach (var instrumentation in enabledInstrumentations.Split(separator: ','))
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:43]: This pattern doesn't come from this PR, but, now we have the typical separator, comma, and the one for plugins hard coded in many places. It seems reasonable to consolidate them. That said it will be better to do it in a separate PR.

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+        /// <summary>
+        /// Gets the list of meters to be added to the MeterProvider at the startup.
+        /// </summary>
+        public IList<string> Meters { get; } = new List<string>();
+
+        internal static MeterSettings FromDefaultSources()
+        {
+            var configurationSource = new CompositeConfigurationSource
+        {
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:129]: Indentation is broken here.

**Review by pjanotti — APPROVED** [review]
LGTM minus the indentation issue.

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs (hunk 1 above)
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:43]: Created issue #678 to track this effort.

**Review by pellared — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                        instrumentations[instrumentation] = parsedType;
+                    }
+                    else
+                    {
+                        throw new ArgumentException($"The \"{instrumentation}\" is not recognized as supported metrics instrumentation and cannot be enabled");
+                    }
+                }
+            }
+
+            var disabledInstrumentations = source.GetString(ConfigurationKeys.Metrics.DisabledInstrumentations);
```
Comment (Kielek) [inline src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:56]: I know that we have exactly same config structure for traces, but I have some doubts that we need it.
For now, it can be replaced by just removing the instrumentation from the enabled instrumentations list. It will change if we enable the metrics by default.

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs (hunk 3 above)
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:56]: I was thinking about the same as well.
But then I thought that it can be used for `Integrations` as well (in the same way as for traces). Nothing for adding byte-code metrics instrumentation, right?

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+#endif
+            };
+
+            return new MeterSettings(configurationSource);
+        }
+
+        private static MetricsExporter ParseMetricExporter(IConfigurationSource source)
+        {
+            var metricsExporterEnvVar = source.GetString(ConfigurationKeys.Metrics.Exporter) ?? "otlp";
+            switch (metricsExporterEnvVar)
```
Comment (zacharycmontoya) [inline src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:144]: Should we do a `string.ToLower()` here just to be safe?

File: src/OpenTelemetry.AutoInstrumentation/Configuration/Settings.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        /// <summary>
+        /// Gets the the OTLP transport protocol. Supported values: Grpc and HttpProtobuf.
+        /// </summary>
+        public OtlpExportProtocol? OtlpExportProtocol { get; }
+
+        /// <summary>
+        /// Gets a collection of <see cref="Integrations"/> keyed by integration name.
+        /// </summary>
+        public IntegrationSettingsCollection Integrations { get; }
```
Comment (zacharycmontoya) [inline src/OpenTelemetry.AutoInstrumentation/Configuration/Settings.cs:53]: The backing collection of integrations is sourced from the `IntegrationRegistry`, which is populated by the values from the `TracerInstrumentation` enum. Did you want to combine the metrics integrations with the tracer integrations, or keep them separate?

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MetricsExporter.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.AutoInstrumentation.Configuration;
+
+/// <summary>
+/// Enum representing supported metrics exporters.
+/// </summary>
+public enum MetricsExporter
```
Comment (zacharycmontoya) [inline src/OpenTelemetry.AutoInstrumentation/Configuration/MetricsExporter.cs:22]: This enum starts with `Metrics` but other types start with `Meter`. Are there OTEL naming guidelines for this? I'm unfamiliar with OTEL metrics so maybe I'm just unfamiliar with the conventions

**Review by zacharycmontoya — APPROVED** [review]
LGTM, just left some questions on smaller details

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs (hunk 3 above)
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:56]: Correct.

Below is the code from `EnvironmentConfigurationTracerHelper.cs`. Instrumentation list is parsed directly and added to `TracerProviderBuilder`. Incase of metrics addition is required to `MeterProviderBuilder`. Instrumentation libraries supported for traces and metrics as different and this needs to be handled separately. OpenTelemetry SDK treats traces, metrics and logs as 3 different signals and provides and option to enable/disable all 3 signals on a need basis. Design motivation is same here to treat all the signals different.

```
    public static TracerProviderBuilder UseEnvironmentVariables(this TracerProviderBuilder builder, TracerSettings settings)
    {
        var resourceBuilder = ResourceBuilder.CreateDefault();

        builder
            .SetResourceBuilder(resourceBuilder)
            .SetExporter(settings);

        foreach (var enabledInstrumentation in settings.EnabledInstrumentations)
        {
            if (AddInstrumentation.TryGetValue(enabledInstrumentation, out var addInstrumentation))
            {
                addInstrumentation(builder);
            }
        }

        builder.AddSource(settings.ActivitySources.ToArray());
        foreach (var legacySource in settings.LegacySources)
        {
            builder.AddLegacySource(legacySource);
        }

        return builder;
    }

```

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs (hunk 4 above)
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:144]: Very good recommendation, we have this pattern uses in TracerSettings and other places. I want to get this covered in a different PR. What are your thoughts.

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MetricsExporter.cs (hunk 6 above)
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MetricsExporter.cs:22]: All metrics API based terminologies uses prefix `Meter` - https://opentelemetry.io/docs/reference/specification/metrics/api/.

Metrics Exporters starts with `Metrics` as per specification - https://opentelemetry.io/docs/reference/specification/metrics/sdk_exporters/otlp/

File: src/OpenTelemetry.AutoInstrumentation/Configuration/Settings.cs (hunk 5 above)
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/Settings.cs:53]: At this point, no integrations are needed for metrics. I want to keep them separate when use case arises for it.

Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/Settings.cs:53]: Moved `IntegrationSettingsCollection` to `TracerSettings`.

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs (hunk 4 above)
Comment (zacharycmontoya) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:144]: Yes another PR is fine with me 😄

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MetricsExporter.cs (hunk 6 above)
Comment (zacharycmontoya) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MetricsExporter.cs:22]: Thanks for the links, that helps a lot 👍🏼

File: src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs (hunk 3 above)
Comment (zacharycmontoya) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:56]: Are there any requested changes here? If not, I think we can merge the PR in its current state

Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/MeterSettings.cs:56]: Similar question has come during Environment variables too. It's better to discuss the topic in next SIG meeting. I will get the PR merged to unblock the metrics integration.
